crates/scmi/src/devices/common.rs

```
@@ -0,0 +1,531 @@
+        self.0.handle(message_id, parameters)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::assert_eq;
+
+    use super::*;
+
+    #[test]
+    fn test_help() {
+        let help = devices_help();
+        assert!(
+            help.contains("Available devices:\n"),
+            "global label missing"
+        );
+        assert!(help.contains("fake:"), "sensor name missing");
+        assert!(
+            help.contains("fake accelerometer"),
+            "short description missing"
+        );
+        assert!(help.contains("3-axes sensor"), "long description missing");
+        assert!(help.contains("Parameters:\n"), "parameter label missing");
+        assert!(help.contains("- name:"), "parameter `name' missing");
+    }
+
+    fn device_properties() -> DeviceProperties {
+        let s = String::from;
+        DeviceProperties::new(vec![
+            (s("foo"), s("val1")),
+            (s("def"), s("val2")),
+            (s("bar"), s("val3")),
+        ])
+    }
+
+    #[test]
+    fn test_device_properties() {
+        let properties = device_properties();
+        assert_eq!(properties.get("bar"), Some("val3"));
+        assert_eq!(properties.get("baz"), None);
+        assert_eq!(
+            properties.names(),
+            HashSet::from_iter(["foo", "def", "bar"].into_iter())
+        );
+        let expected = ["abc", "def", "ghi"];
+        let missing = properties.missing(&expected);
+        assert_eq!(missing, HashSet::from_iter(["abc", "ghi"].into_iter()));
+        let extra = properties.extra(&expected);
+        assert_eq!(extra, HashSet::from_iter(["foo", "bar"].into_iter()));
+    }
+
+    #[test]
+    fn test_check_device_properties() {
+        let properties = device_properties();
+        let result = check_device_properties(&properties, &["abc", "def", "ghi"], &["foo", "baz"]);
+        assert_eq!(result, Err("Missing device parameters: abc, ghi".into()));
```

This sound like the check_device_properties should return a proper Error type instead of a string but no strong opinion.

---

I agree, done.